[kaungzinye] iP

kaungzinye opened the pull request:
# **Gojo Satoru - The Strongest Taskmaster 🤖**  

Gojo Satoru, the **undisputed** strongest sorcerer, has descended from Jujutsu High to **effortlessly** manage your tasks! When he’s not flexing his Infinity technique or mocking his enemies, he’s here to **make your life easier**.  

> "Nah I'd Win." — Gojo, before he lost.  

## **Features Gojo Brings to the Table** 🏆  

Here's what **the honored one** can do for you:  

- 📋 **Maintain a to-do list** so you won’t forget things.  
- ⏳ **Track deadlines** for important tasks before they **disappear into the Unlimited Void**.  
- 📅 **Keep notes on upcoming events** because the time between you and your deadlines aren't Infinite.  
- 🔍 **Search for tasks** that contain a certain word or phrase—like **"You're weak"**.  
- ⏰ **Tell you what's due on a specific day** because Gojo knows the future (kind of).  

---

## **Inside Gojo’s Brain (Code Structure) 🧠**  

Gojo doesn’t just rely on his eyes—he's got a solid **object-oriented** core. Here’s how he’s built:  

```java
package gojosatoru;


import java.time.format.DateTimeFormatter;
import java.util.Objects;
import java.util.Scanner;

import gojosatoru.command.Command;
import gojosatoru.exceptions.GojoException;
import gojosatoru.parser.Parser;
import gojosatoru.storage.Storage;
import gojosatoru.tasks.TaskList;
import gojosatoru.ui.Ui;

public class GojoSatoru {
    private static final String FILE_PATH = "./src/main/data/cursedEnergy.txt";
    private static final Ui UI = new Ui();

    private static String inputDateFormat = "dd/MM/yyyy HHmm";
    private static String outputDateFormat = "MMM dd yyyy HHmm";
    private static DateTimeFormatter inputFormatter = DateTimeFormatter.ofPattern(inputDateFormat);
    private static DateTimeFormatter outputFormatter = DateTimeFormatter.ofPattern(outputDateFormat);
    private static Command command = new Command(inputFormatter, outputFormatter, inputDateFormat, UI);
    private static Storage storage = new Storage(FILE_PATH, command, inputFormatter, outputFormatter);
    private static Parser parser = new Parser(command);
```

💡 Gojo follows **object-oriented programming (OOP) principles**—a structured way of thinking about **objects, interactions, and behaviors**. You can learn more about OOP [[here](https://nus-cs2103-ay2425s2.github.io/website/schedule/week1/topics.html#W1-1)](https://nus-cs2103-ay2425s2.github.io/website/schedule/week1/topics.html#W1-1).  

---

## **Working Commands** 🎤  

Because Gojo understands **simplicity is key** (because his Six Eyes complicates everything), here’s how you can **talk to him**:  

- 📝 `list` — Displays all your tasks.  
- ✅ `todo` `<task>` — Adds a new **to-do** task.  
- ⏳ `deadline` `<task>` `/by` `<date>` — Adds a **deadline**.  
- 🎭 `event` `<event>` `/from` `<start>` `/to` `<finish>` — Adds an **event**.  
- 🔎 `find` `<keyword>` — Searches for tasks.  
- 👋 `bye` — Ends the session (Gojo never says goodbye, though).  

---

## **What Gojo Will Learn Next** ✨  

Like his **domain expansion**, Gojo's abilities are still evolving. Here’s what he’s working on:  

- [x] **Text UI** (Of course, he's already mastered this)  
- [ ] **Search Function** (He's a genius, so it'll be easy)  
- [ ] **Graphical UI** (Can’t fight Sukuna with just text, right?)  
- [ ] **Interactive Elements** (Gojo’s personality is interactive by default)  
- [ ] **Conversational Ability** (Sarcastic responses included 😏)  

---

## **How to Summon Gojo Satoru** 🌀  

To test Gojo’s abilities, follow these **simple** steps:  

```bash
./gradlew test
```

For more detailed test output, including **passed, skipped, and failed tests**, use:  

```bash
./gradlew test --info
```

If you need **full debug logs**, run:  

```bash
./gradlew test --debug
```

---

## **Final Notes** 📝  

> "Hollow Purple? Nah, this PR is already overpowered enough." — Gojo  

This PR includes various **improvements and bug fixes**, ensuring Gojo **stays at the top**.  

For more markdown tricks, check out [[GitHub Flavored Markdown](https://guides.github.com/features/mastering-markdown/)](https://guides.github.com/features/mastering-markdown/).  

Oh, and one last thing:  

**Bold moves**, *stylish formatting*, and ~~erased opponents~~ are all included. 😎  

---

### **Ready to merge? Or do you need Six Eyes to review it first?** 🔵

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

Commit:
Finished level 0

Commit:
Please enter the commit message for your changes. Lines starting

Commit:
LEvel 2

Commit:
level-3

Commit:
level-3 fix

Commit:
Level-4

Commit:
UI testing

Commit:
Level 7: Save tasks to disk and load on startup

This commit implements the following features and improvements:

- **Save tasks to disk**: The task list is automatically saved to the hard disk whenever it changes. This ensures that the tasks are persisted between sessions.
- **Load tasks from disk**: The tasks are loaded from the hard disk when the chatbot starts up. This allows the chatbot to resume from where it left off.
- **File path handling**: The file paths are specified using relative paths and in an OS-independent way to ensure compatibility across different operating systems.
- **Folder and file creation**: The code handles the case where the data file or the required folder does not exist at the start. This ensures that the chatbot can be run on a new computer without issues.
- **Data file corruption handling**: The code includes basic handling for the situation where the data file is corrupted (i.e., content not in the expected format) by throwing IO errors.

Additionally, the project structure has been restructured to better organize the code and resources.

Example of the data file format as given in the task requirements:
T | 1 | read book
D | 0 | return book | June 6th
E | 0 | project meeting | Aug 6th 2-4pm
T | 1 | join sports club

This change improves the stability and usability of the application by ensuring that tasks are persisted and loaded correctly.

Commit:
Merge branch 'branch-level-7'
Merge branch 'branch-Level-7' into master

This merge integrates the changes from `branch-Level-7` into the `master` branch. The changes include:

- Saving tasks to disk and loading them on startup.
- Handling file paths in an OS-independent way.
- Creating necessary folders and files if they do not exist.
- Basic handling for corrupted data files.
- Restructuring the project folder for better organization.

This merge ensures that the `master` branch has the latest features and improvements from `branch-Level-7`.

Commit:
Implement Level 8: Dates and Times, and Automated Testing

- Teach the chatbot to understand dates and times for deadlines and events.
- Store deadline dates as `java.time.LocalDateTime` in task objects.
- Accept dates in `yyyy-MM-dd'T'HH:mm` format and print in `MMM dd yyyy HH:mm` format.
- Add a command to print deadlines/events occurring on a specific date.
- Implement a handler class to manage task creation functions.
- Add automated text UI testing using I/O redirection technique.
- Test functionalities including creating, marking, unmarking, deleting tasks, and handling invalid dates.

Commit:
Implement Level 8: Dates and Times, and Automated Testing

- Teach the chatbot to understand dates and times for deadlines and events.
- Store deadline dates as `java.time.LocalDateTime` in task objects.
- Accept dates in `yyyy-MM-dd'T'HH:mm` format and print in `MMM dd yyyy HH:mm` format.
- Add a command to print deadlines/events occurring on a specific date.
- Implement a handler class to manage task creation functions.
- Add automated text UI testing using I/O redirection technique.
- Test functionalities including creating, marking, unmarking, deleting tasks, and handling invalid dates.

Commit:
Merge branch-Level-8

Commit:
Refactor: Extract task management logic into TaskList class

- Created TaskList class to handle addition, deletion, and retrieval of tasks.
- Updated Storage class to use TaskList for task operations.
- Modified Parser class to work with TaskList instead of ArrayList<Task>.
- Updated GojoSatoru class to initialize and use TaskList.

Commit:
Merge branch 'A-MoreOOP'
Merge

Commit:
Merge remote-tracking branch 'origin/add-gradle-support'
Merging add-gradle-support into master branch

Commit:
Added Junit Tests

kubrian commented on src/main/java/gojosatoru/tasks/Event.java:
```diff
@@ -0,0 +1,31 @@
+    }
+    return output;
+  }
+
```
Consider the following
* Addition of whitespaces around conditions and keywords
* Addition of linebreaks
* Using Egyptian style brackets

Noticed the same issue in other places too.

kubrian reviewed (commented):
I like the naming of variables and methods, it is rather intuitive.

I would suggest a stricter adherence to the style guide, in particular for these two points:
* Write descriptive header comments for all public classes/methods
* Basic indentation should be 4 spaces

DanielJames0302 commented on src/main/java/gojosatoru/exceptions/GojoException.java:
```diff
@@ -0,0 +1,17 @@
+import gojosatoru.ui.Ui;
+
+public class GojoException extends Exception {
+  private static final Ui UI = new Ui();
```
Any reason why you had the UI in this class instead of putting it in the Gojo class ?

kaungzinye replied:
Intially wanted to format the output of the exceptions using this class but it caused indentation errors so I  did not use it, will change implementation in future commits.

DanielJames0302 commented on src/main/java/gojosatoru/tasks/Task.java:
```diff
@@ -0,0 +1,29 @@
+  // Method to be implemented to save the task in a specific format
+  public abstract String toSaveFormat();
+}
+
```
I suggest adding an abstract function to execute the task so that you can specify how a particular task can be executed in the subclasses of the Task class

DanielJames0302 commented on src/main/java/gojosatoru/tasks/Deadline.java:
```diff
@@ -0,0 +1,28 @@
+    else{
+      output = "[D][ ] " + taskDescription + " (by: " + outputFormatter.format(by) + ")";
+    }
+    return output;
```
Shoud this part use ternary operator on `this.completed` to prevent duplicated code in the if-else branches ? I like how you do it at line 26

DanielJames0302 reviewed (commented):
LGTM. Just some nits to fix.

Commit:
A-Jar, added jar creation functionality

Commit:
Add header comments to non-private classes and methods

- Added header comments to non-private classes and methods to all classes

Commit:
Refactor code to adhere to coding standards

- Renamed  to  to comply with naming conventions.
- Fixed import order and grouping according to  rule.
- Ensured all imports are used and necessary.

Commit:
Merge branch

Commit:
Coding Standards

Commit:
Merge branch 'branch-A-CodingStandards'

merge

Commit:
Implemented level-9 Find

Commit:
Manual merge with conflict resolution

Commit:
redoing checkstyle

Commit:
Added back Checkstyle

Commit:
Merge branch 'branch-A-Checkstyle'
chore: integrate CheckStyle for coding standards

- Added CheckStyle configuration to enforce coding standards.
- Referenced the SE-EDU guide for CheckStyle setup.
- Ensured the project adheres to the specified coding conventions.

jensenhuangyankai commented on src/main/java/gojosatoru/command/Command.java:
```diff
@@ -0,0 +1,297 @@
+     * @throws MissingArgumentException if the task description, start time, or end time is missing
+     * @throws InvalidDateException if the start time is not before the end time
+     */
+    public Event handleEvents(String input) throws MissingArgumentException, InvalidDateException {
```
Good job following the coding standard for the naming of methods, keep it up

jensenhuangyankai commented on src/main/java/gojosatoru/storage/Storage.java:
```diff
@@ -0,0 +1,152 @@
+                Task task;
+                try {
+                    switch (taskDetails[0]) {
+                    case "T":
```
Nothing wrong with the switch statement here, but consider using a lambda switch statement. Might prevent fall-through of cases.

Commit:
feat: add JavaFX GUI to chatbot and update tests

- Implement JavaFX GUI for the chatbot application
- Update  method to check for missing start and end times
- Fix test cases to match the expected error messages
- Correct expected message in
- Ensure correct error message for invalid task indices
- Adjust  to match actual output
- Update  with additional test cases for GUI-related changes

Commit:
Merge branch 'branch-Level-10'
Merge branch 'feature/javafx-gui' into master

- Implement JavaFX GUI for the chatbot application
- Update `handleEvents` method to check for missing start and end times
- Fix test cases to match the expected error messages
- Correct expected message in `testTaskNotFoundShowsError`
- Ensure correct error message for invalid task indices
- Adjust `testSpecialCharacterTask` to match actual output
- Update `MainTest` with additional test cases for GUI-related changes

Commit:
Adding VarArgs by refactoring handleFind and findTasks; update unit tests

The handleFind and findTasks functions now implement varargs.

Additionally, unit tests have been updated to align with these changes, ensuring comprehensive test coverage and validating the refactored functions' behavior.

This refactoring enhances the codebase's modularity and simplifies future updates or feature additions.

Commit:
MainTest.java: fix Checkstyle violations

Reordered imports to follow Checkstyle's CustomImportOrder rules.
Wrapped long lines exceeding 120 characters to comply with line length restrictions.

These changes ensure code adheres to the project's style guidelines, improving readability and maintainability.

No functional changes were made.

Commit:
A-Assertions: add assertions and fix test directory

Added assertions in MainTest.java and other chatbot files to ensure correctness.

Fixed MainTest.java placement by adding it to the testdirectory folder, ensuring proper test organization.

kaungzinye committed:
Merge pull request #3 from kaungzinye/branch-A-Assertions

 A-Assertions Add Assertions & Fix Test Directory Structure

Commit:
MissingArgumentException: remove repeated strings

Refactored MissingArgumentException to eliminate repeated strings, improving code readability and maintainability.

kaungzinye committed:
Merge pull request #4 from kaungzinye/branch-A-CodeQuality

A-CodingQuality

Commit:
Add ci.yml for CI/CD integration

Created a ci.yml file to set up continuous integration and deployment.

This ensures automated testing and builds for the repository, improving development workflow and code reliability.

Commit:
Fix CI/CD: Ensure cursedEnergy.txt exists before tests

Commit:
Fix CI/CD: Ensure javaFX exists before tests

Commit:
C-DetectDuplicates: Handle duplicate tasks

Add functionality to detect and manage duplicate tasks before adding them.
This prevents multiple identical tasks from being added to the task list.

Changes:
- Added `findDuplicate` method in `TaskList` to check for existing tasks.
- Implemented `handleDuplicate` in `Command` to prompt users before adding duplicates.
- Updated `setNextInput` in `Ui` to improve command parsing for test cases.

This enhancement improves task list integrity and ensures users are notified of duplicate entries.

Commit:
Fix JavaFX setup for GitHub Actions

Added JavaFX installation steps for macOS and Linux.
Ensured cursedEnergy.txt file exists to prevent missing file errors.
Updated workflow to properly handle different OS environments.

Commit:
Add More Personality To Bot-m

Commit:
Update README

Commit:
Update README

Commit:
Add Jar

Commit:
Add formatting

Commit:
Add Final Jar

Commit:
Add Final Jar

Commit:
Update Ui.png

Commit:
Add Jar

Commit:
Add Launcher workaround
